api/server/router/container/container_routes.go

```diff
@@ -475,6 +475,13 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 		}
 	}
 
+	// When using API older than 1.32, the client is not able to set
+	// --ipc shareable but expects it to be the default, so let's
+	// explicitly set it here, overriding engine's default-ipc-mode.
+	if hostConfig != nil && hostConfig.IpcMode.IsEmpty() && versions.LessThan(version, "1.32") {
+		hostConfig.IpcMode = container.IpcMode("shareable")
```

Comment:
Sadly this API version is a bit out of date.

Reply:
> Sadly this API version is a bit out of date.

￼You mean for the case of k8s client? Originally this patch was intended for compatibility with older docker cli which was not able to set "shareable".

It looks like kubernetes is currently using docker api 1.38. The problem is, we can't have separate cases for docker-cli, kubelet client and others.

I can't think of any easy solution here, rather than wait a few years and assume the fix (https://github.com/kubernetes/kubernetes/pull/70826) has made its way to all k8s versions, then change the default to be private.

Reply:
Shouldn't this just be gated to the current API version?

Reply:
I don't think so. The meaning of this hunk is to use old default of "shareable" for older clients (i.e. those that have no way to set "shareable" ipcmode explicitly). As the functionality to set shareable ipcmode appeared in API v1.32, the check is for < 1.32.

Reply:
Sorry, I don't think we should change the default without gating it by the api version where we've changed the default.

Reply:
OK, so we have a server (daemon, engine) and a client, and an API version.

Since API 1.32
1. the client can explicitly set shareable IPC mode;
2. the server can have default IPC mode set to shareable (using `default-ipc-mode` in either daemon.json, or as a command line option).

The code in question was added as a client with API < 1.32 is unable to set shareable IPC mode at all, not because the default is changed. Strictly speaking, it does not belong to this PR, but is rather a followup to #34087.

As far as I understand, if we change `1.32` in the code above to a later version, the above functionality (2) of setting default IPC mode will cease to work.

Reply:
My point is, changing the default should be gated by the API version we are introducing the change in rather than some old API version. So basically we shouldn't send an empty value for any version < the current version and it should be the old default ("shareable").

Reply:
We seem to have a disconnect here:

< 1.32 there was no option and only shared was used
>= 1.32 default is shared but client can change

My proposal is to say if version is >= 1.32 < 1.40 (or whatever the current released API version is at this point...) to keep the default as shareable, and for the upcoming release, daemon default of private.

Reply:
Yes, I'm on @cpuguy83's side of this; if we change this value for API v1.32..v1.40 in the upcoming release, it means that users that are _currently_ using API v1.32..v1.40 that update the daemon will all of a sudden get a different behavior on the same API version.

Basically, if a user pins to API version x, then updating the daemon to whatever newer version should maintain exactly the same behavior (baring bug-fixes of course). If they do _not_ pin to a specific version (or explicitly update to pin to a newer API version) they are ok with getting new behavior introduced by the newer API version.

Reply:
Of course, we can recommend somewhere that users should configure their daemon to;

- use `private` or `shareable` as a default if they're using API v1.32 or up if they do not require `shared`
- or (the reverse); configure their daemon to use `shared` as default before updating to the upcoming release that has this change (if they want to use API v1.40+ but need to keep the old behavior) (this one should definitely go into the release notes)

Reply:
Perhaps I made a very wrong decision to put these two commits together in one PR, and this is the source of confusion. Let me remove the second commit and put it into a separate PR, hoping it will be easier to grasp that way.

Reply:
#38700